source/opt/instrument_pass.cpp

```diff
@@ -0,0 +1,951 @@
+    uint32_t obufTyId = type_mgr->GetTypeInstruction(reg_obuf_ty);
+    AddDecoration(obufTyId, SpvDecorationBlock);
+    AddMemberDecoration(obufTyId, kDebugOutputSizeOffset, SpvDecorationOffset,
+        0);
+    AddMemberDecoration(obufTyId, kDebugOutputDataOffset, SpvDecorationOffset,
+        4);
+    uint32_t obufTyPtrId_ = type_mgr->FindPointerToType(obufTyId,
+        SpvStorageClassStorageBuffer);
+    output_buffer_id_ = TakeNextId();
+    std::unique_ptr<Instruction> newVarOp(
+        new Instruction(context(), SpvOpVariable, obufTyPtrId_,
+            output_buffer_id_,
+            { { spv_operand_type_t::SPV_OPERAND_TYPE_LITERAL_INTEGER,
+                { SpvStorageClassStorageBuffer } } }));
+    get_def_use_mgr()->AnalyzeInstDefUse(&*newVarOp);
+    get_module()->AddGlobalValue(std::move(newVarOp));
+    AddDecorationVal(output_buffer_id_, SpvDecorationDescriptorSet, desc_set_);
+    AddDecorationVal(output_buffer_id_, SpvDecorationBinding,
+        GetOutputBufferBinding());
+  }
+  return output_buffer_id_;
+}
+
+uint32_t InstrumentPass::GetVec4FloatId() {
+  if (v4float_id_ == 0) {
+    analysis::TypeManager* type_mgr = context()->get_type_mgr();
+    analysis::Float float_ty(32);
+    analysis::Type* reg_float_ty = type_mgr->GetRegisteredType(&float_ty);
+    analysis::Vector v4float_ty(reg_float_ty, 4);
+    analysis::Type* reg_v4float_ty = type_mgr->GetRegisteredType(&v4float_ty);
+    v4float_id_ = type_mgr->GetTypeInstruction(reg_v4float_ty);
+  }
+  return v4float_id_;
+}
+
+uint32_t InstrumentPass::GetUintId() {
+  if (uint_id_ == 0) {
+    analysis::TypeManager* type_mgr = context()->get_type_mgr();
+    analysis::Integer uint_ty(32, false);
+    analysis::Type* reg_uint_ty = type_mgr->GetRegisteredType(&uint_ty);
+    uint_id_ = type_mgr->GetTypeInstruction(reg_uint_ty);
+  }
+  return uint_id_;
+}
+
+uint32_t InstrumentPass::GetVec4UintId() {
+  if (v4uint_id_ == 0) {
+    analysis::TypeManager* type_mgr = context()->get_type_mgr();
+    analysis::Integer uint_ty(32, false);
+    analysis::Type* reg_uint_ty = type_mgr->GetRegisteredType(&uint_ty);
+    analysis::Vector v4uint_ty(reg_uint_ty, 4);
+    analysis::Type* reg_v4uint_ty = type_mgr->GetRegisteredType(&v4uint_ty);
+    v4uint_id_ = type_mgr->GetTypeInstruction(reg_v4uint_ty);
+  }
+  return v4uint_id_;
+}
+
+uint32_t InstrumentPass::GetBoolId() {
+  if (bool_id_ == 0) {
+    analysis::TypeManager* type_mgr = context()->get_type_mgr();
+    analysis::Bool bool_ty;
+    analysis::Type* reg_bool_ty = type_mgr->GetRegisteredType(&bool_ty);
+    bool_id_ = type_mgr->GetTypeInstruction(reg_bool_ty);
+  }
+  return bool_id_;
+}
+
+uint32_t InstrumentPass::GetVoidId() {
+  analysis::TypeManager* type_mgr = context()->get_type_mgr();
+  analysis::Void void_ty;
+  analysis::Type* reg_void_ty = type_mgr->GetRegisteredType(&void_ty);
+  return type_mgr->GetTypeInstruction(reg_void_ty);
+}
+
+uint32_t InstrumentPass::GetBuiltinVarId(uint32_t builtin, uint32_t type_id,
```

Comment:
I cut down the operands and moved into context and created a new analysis for it.

I have also moved some decoration generating code into the decoration manager and some constant generation code into the constant manager.

I think I have moved nearly all code that might be generally useful into areas where it can be shared.